calculator_test.go

```
@@ -2,23 +2,167 @@ package calculator_test
+		want := a + b
+		got := calculator.Add(a, b)
+		if want != got {
+			t.Errorf("random test: want %f, got %f", want, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+	for _, tc := range testCasesSubtract {
+		want := tc.want
+		got := calculator.Subtract(tc.a, tc.b)
+		if want != got {
+			t.Errorf("test %s: want %f, got %f", tc.comment, want, got)
+		}
+	}
+}
+
+func TestSubtractRandom(t *testing.T) {
+	t.Parallel()
+	for i := 0; i < 100; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a - b
+		got := calculator.Subtract(a, b)
+		if want != got {
+			t.Errorf("random test: want %f, got %f", want, got)
+		}
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestMultiply(t *testing.T) {
+	t.Parallel()
+	for _, tc := range testCasesMultiply {
+		want := tc.want
+		got := calculator.Multiply(tc.a, tc.b)
+		if want != got {
+			t.Errorf("test %s: want %f, got %f", tc.comment, want, got)
+		}
+	}
+}
+
+func TestMultiplyRandom(t *testing.T) {
+	t.Parallel()
+	for i := 0; i < 100; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a * b
+		got := calculator.Multiply(a, b)
+		if want != got {
+			t.Errorf("random test: want %f, got %f", want, got)
+		}
+	}
+}
+
+func TestDivide(t *testing.T) {
+	t.Parallel()
+	for _, tc := range testCasesDivide {
+		want := tc.want
+		got, err := calculator.Divide(tc.a, tc.b)
+		errorReceived := err != nil
+		if errorReceived != tc.errExpected {
+			t.Fatalf("test %s: expected err %t, got %t", tc.comment, tc.errExpected, errorReceived)
+		}
+		if want != got {
+			t.Errorf("test %s: want %f, got %f", tc.comment, want, got)
+		}
+	}
+}
+
+func TestDivideRandom(t *testing.T) {
+	t.Parallel()
+	for i := 0; i < 100; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a / b
+		got, _ := calculator.Divide(a, b)
```

We never ignore errors! As we've just seen, we _could_ get an error here, so we need to check it.